lib/bpf/helper.go

```diff
@@ -19,66 +19,104 @@ limitations under the License.
-func attachProbe(module *bcc.Module, eventName string, functionName string) error {
-	kprobe, err := module.LoadKprobe(functionName)
+func LoadBPF(code []byte, name string) (*libbpfgo.Module, error) {
+	mod, err := libbpfgo.NewModuleFromBuffer(code, "hook")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Load into the kernel
+	if err = mod.BPFLoadObject(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return mod, nil
+}
+
+// attaches both kprobe and kretprobe
+func AttachKprobe(mod *libbpfgo.Module, name string) error {
+	prog, err := mod.GetProgram("kprobe__" + name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	_, err = prog.AttachKprobe(name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	prog, err = mod.GetProgram("kretprobe__" + name)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	err = module.AttachKprobe(eventName, kprobe, -1)
+	_, err = prog.AttachKretprobe(name)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
 	return nil
 }
 
-// attachRetProbe will attach a kretprobe to the given function name.
-func attachRetProbe(module *bcc.Module, eventName string, functionName string) error {
-	kretprobe, err := module.LoadKprobe(functionName)
+func AttachTracepoint(mod *libbpfgo.Module, category, name string) error {
+	prog, err := mod.GetProgram("tracepoint__" + category + "__" + name)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	err = module.AttachKretprobe(eventName, kretprobe, -1)
-	if err != nil {
+	_, err = prog.AttachTracepoint(category + ":" + name)
+	return err
+}
+
+func AttachSyscallTracepoint(mod *libbpfgo.Module, syscall string) error {
+	if err := AttachTracepoint(mod, "syscalls", "sys_enter_" + syscall); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := AttachTracepoint(mod, "syscalls", "sys_exit_" + syscall); err != nil {
 		return trace.Wrap(err)
 	}
 
 	return nil
 }
 
-// openPerfBuffer will open a perf buffer for a particular module.
-func openPerfBuffer(module *bcc.Module, perfMaps []*bcc.PerfMap, pageCount int, name string) (<-chan []byte, <-chan uint64, error) {
-	var err error
-
-	eventCh := make(chan []byte, chanSize)
-	lostCh := make(chan uint64, chanSize)
+type RingBuffer struct {
+	buf *libbpfgo.RingBuffer
+	EventCh chan []byte
+}
 
-	table := bcc.NewTable(module.TableId(name), module)
+// openPerfBuffer will open a perf buffer for a particular module.
+func NewRingBuffer(mod *libbpfgo.Module, name string) (*RingBuffer, error) {
+	rb := &RingBuffer{
+		EventCh: make(chan []byte, chanSize),
+	}
 
-	perfMap, err := bcc.InitPerfMap(table, eventCh, lostCh, uint(pageCount))
+	var err error
+	rb.buf, err = mod.InitRingBuf(name, rb.EventCh)
 	if err != nil {
-		return nil, nil, trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
-	perfMap.Start()
 
-	perfMaps = append(perfMaps, perfMap)
+	rb.buf.Start()
+
+	return rb, nil
+}
 
-	return eventCh, lostCh, nil
+func (rb *RingBuffer) Stop() {
+	rb.buf.Stop()
+	// don't Close rb.buf as it'll be closed as part of Module.Close()
```

Comment:
Move this comment to function description, so:

```go
// Stop will stop the ring buffer. Note that the underlying ring buffer is not
// closed because it's closed as part of Module.Close().
```